### DISCUSSION TITLE
how can I avoid a cyclic import here

### DISCUSSION THREAD

**Discussion by georgfaust** [cue-lang/cue#1942] [Q&A]
In my main.cue I build a some nested structure

```
package main

import "Export export" // cyclic

#foo: bar: baz: ...

foo: export.foo
```

I have some 'functions' that extract data in various formats using comprehensions and `list.FlattenN`.

```
package export

import "main"  // cyclic

foo: list.FlattenN([for f in #foo...])
```

I would like to put those into a module, but I would need to import `#foo` from main and `export` into main, which is a cyclic import and thus hangs. How can this be solved?

**Comment by verdverm** [comment]
The semantics here for CUE are very close to Go's, so a few points

- If it "hangs" rather than exiting with a cycle or other error, that is a bug
- modules require a domain in the first part and imports must be qualified with their module, so what you have above is not valid CUE imports
- you can use the "function" pattern ( https://cuetorials.com/patterns/functions/ ) to build definitions you can import and use with other values

**Comment by eadlam** [comment]
@georgfaust

**export.cue**
```cue
package export

import "list"

#Flatten: {
    #data: {...}    
    #n: int | *1
    list.FlattenN([for k, v in #data {v}], #n)
}
```

**main.cue**
```cue
package main

import "example.com:export"

#foo: {
    bar: [1, [2]] 
    baz: [[3, 4]]
}

foo: {export.#Flatten, #data: #foo, #n: 2}
```

**cli**
```txt
$ cue export example.com:main
{
    "foo": [
        1,
        2,
        3,
        4
    ]
}
```
